### DISCUSSION TITLE
teleport-cluster helm chart insecureSkipProxyTLSVerify

### DISCUSSION THREAD

**Discussion by mcandio** [gravitational/teleport#11519] [Q&A] [answered]
### Description
The flag --insecure passed when insecureSkipProxyTLSVerify is true is not working

**What happened**:
I'm using teleport with aws NLB and I want to use --insecure-no-tls flag.
In teleport-cluster helm chart it is defined as --insecure when set insecureSkipProxyTLSVerify to either true or false.

Fixed this by adding the arg --insecure-no-tls in a running deployment.

**What you expected to happen**:

### Reproduction Steps
As minimally and precisely as possible, describe step-by-step how to reproduce the problem.
I'm using Helmfile to install the chart. so setting this as the helmfile:
````
---
repositories:
  - name: teleport
    url: https://charts.releases.teleport.dev

helmDefaults:
  wait: true
  timeout: 120
  atomic: false
  createNamespace: true
  kubeContext: arn:aws:eks:us-east-1:773848629605:cluster/devops-develop-eks-cluster
releases:
  - name: teleport
    namespace: teleport
    chart: teleport/teleport-cluster
    values:
      - ./values.yaml
````
and this as the values:
````
insecureSkipProxyTLSVerify: true
````
along with the basic aws configuration and a service annotation with the acm certificate arn

creates a deployment like this one:

````
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "3"
    meta.helm.sh/release-name: teleport
    meta.helm.sh/release-namespace: teleport
  creationTimestamp: "2022-03-23T23:37:34Z"
  generation: 3
  labels:
    app: teleport
    app.kubernetes.io/managed-by: Helm
  name: teleport
  namespace: teleport
  resourceVersion: "53496971"
  uid: b3928d9c-285c-4541-abdd-1189c5e03ec3
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: teleport
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        checksum/config: dda293fcfc6456b6374afa988bda10ebf9d83b031192128c44e4c9eb3dd4a675
      creationTimestamp: null
      labels:
        app: teleport
    spec:
      containers:
      - args:
        - --diag-addr=0.0.0.0:3000
        - --insecure
        image: quay.io/gravitational/teleport:9.0.1
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 6
          httpGet:
            path: /healthz
            port: diag
            scheme: HTTP
          initialDelaySeconds: 5
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 1
        name: teleport
        ports:
        - containerPort: 3000
          name: diag
          protocol: TCP
        readinessProbe:
          failureThreshold: 12
          httpGet:
            path: /readyz
            port: diag
            scheme: HTTP
          initialDelaySeconds: 5
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 1
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/teleport
          name: config
          readOnly: true
        - mountPath: /var/lib/teleport
          name: data
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: teleport
      serviceAccountName: teleport
      terminationGracePeriodSeconds: 30
      volumes:
      - name: gcp-credentials
        secret:
          defaultMode: 420
          secretName: teleport-gcp-credentials
      - configMap:
          defaultMode: 420
          name: teleport
        name: config
      - emptyDir: {}
        name: data
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2022-03-23T23:37:41Z"
    lastUpdateTime: "2022-03-23T23:37:41Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2022-03-23T23:37:34Z"
    lastUpdateTime: "2022-03-24T00:07:59Z"
    message: ReplicaSet "teleport-76fd4df5b5" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 3
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
````
editing the deployment to change the flag from --insecure to --insecure-no-tls solves the empty response and the following error:
````
"\nERROR REPORT:\nOriginal Error: *net.OpError read tcp 10.20.100.236:3080-&gt;10.20.121.199:7355: i/o timeout\nStack Trace:\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:277 github.com/gravitational/teleport/lib/multiplexer.detect\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:233 github.com/gravitational/teleport/lib/multiplexer.(*Mux).detectAndForward\n\t/opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit\nUser Message: failed to peek connection\n\tread tcp 10.20.100.236:3080-&gt;10.20.121.199:7355: i/o timeout" multiplexer/multiplexer.go:236
````

2.
3.

### Server Details

- Where are you running Teleport? AWS

**Comment by webvictim** [comment, marked as answer]
The `insecureSkipProxyTLSVerify` value isn't implemented for the `teleport-cluster` chart - [only for the `teleport-kube-agent` chart](https://github.com/gravitational/teleport/blob/master/examples/chart/teleport-kube-agent/values.yaml#L64-L66).

The `--insecure` flag tells Teleport not to validate TLS certificates when it makes outgoing connections - this can be used to allow Teleport agents to connect to clusters which aren't presenting fully trusted TLS certificates. 

If you do need to use the `--insecure-no-tls` flag (which isn't recommended) then the intent is that you pass this to the `teleport-cluster` chart using [`extraArgs`](https://github.com/gravitational/teleport/blob/master/examples/chart/teleport-cluster/values.yaml#L227-L228).

If you want to use ACM with an AWS NLB, you don't actually need to use `--insecure-no-tls`. See the instructions here: https://goteleport.com/docs/ver/8.0/kubernetes-access/helm/guides/aws/#step-4b7-configure-teleport-to-use-acm-to-handle-tls

**Comment by mcandio** [comment]
Thanks a lot for taking the time to explain how to correctly use this and I'm sorry for not being able to find the documentation referring to this by myself.

**Comment by webvictim** [comment]
That's no problem at all. The docs are actually only in version 8.0 of Teleport at the moment - the PRs merging them to v9 (the current "live" docs) and v10 haven't been merged yet so it's totally understandable that you couldn't find them!

The Helm chart reference is here if it helps: https://goteleport.com/docs/kubernetes-access/helm/reference/#insecureskipproxytlsverify

**Reply by mcandio** [reply]
good!  I've just implemented a cluster with +20 servers and it's working very well with eks acting as proxy/auth.
